File: backend/app/models/complaint.py
import uuid
from datetime import datetime, timezone
from typing import Optional, List
from sqlalchemy import String, Boolean, DateTime, ForeignKey, Float, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database.base_class import Base
from app.constants.enums import ComplaintStatus

class Complaint(Base):
    __tablename__ = "complaints"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    category_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("complaint_categories.id"), nullable=False, index=True)
    municipality_id: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("municipalities.id"), nullable=True, index=True)
    
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[str] = mapped_column(String(50), default=ComplaintStatus.DRAFT.value, nullable=False, index=True)
    severity: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    severity_score: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    image_severity_score: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    ai_confidence_score: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    survey_score: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    weather_score: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    density_score: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    severity_breakdown: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    image_analysis_summary: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    location_name: Mapped[str] = mapped_column(String(500), nullable=False)
    latitude: Mapped[float] = mapped_column(Float, nullable=False)
    longitude: Mapped[float] = mapped_column(Float, nullable=False)
    geo_point: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    assigned_department: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    assigned_officer: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)

    # Soft delete fields
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    deleted_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    deleted_by: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("users.id"), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
        index=True
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False
    )

    # Relationships
    user: Mapped["User"] = relationship("User", foreign_keys=[user_id])
    category: Mapped["ComplaintCategory"] = relationship("ComplaintCategory")
    municipality: Mapped[Optional["Municipality"]] = relationship("Municipality")
    
    attachments: Mapped[List["ComplaintAttachment"]] = relationship(
        "ComplaintAttachment", back_populates="complaint", cascade="all, delete-orphan"
    )
    timeline: Mapped[List["ComplaintStatusHistory"]] = relationship(
        "ComplaintStatusHistory",
        back_populates="complaint",
        cascade="all, delete-orphan",
        order_by="ComplaintStatusHistory.created_at"
    )
    resolution: Mapped[Optional["ResolutionReport"]] = relationship(
        "ResolutionReport", back_populates="complaint", uselist=False, cascade="all, delete-orphan"
    )

File: frontend/types/complaint.ts
import { CategoryResponse } from "./config";

export interface MunicipalityResponse {
  id: string;
  name: string;
  district: string;
  state: string;
  contact_email?: string;
}

export interface TimelineEventResponse {
  id: string;
  status: string;
  remarks: string;
  created_at: string;
}

export interface AttachmentResponse {
  id: string;
  public_url: string;
  file_type: string;
  file_name: string;
  file_size_bytes: number;
}

export interface ResolutionResponse {
  id: string;
  summary: string;
  department: string;
  officer_name: string;
  actions: string;
  before_image_url?: string;
  after_image_url?: string;
  date_resolved: string;
}

export interface ComplaintDetailResponse {
  id: string;
  title: string;
  description: string;
  status: string;
  severity?: string;
  severity_score?: number;
  image_severity_score?: number;
  ai_confidence_score?: number;
  survey_score?: number;
  weather_score?: number;
  density_score?: number;
  severity_breakdown?: string;
  image_analysis_summary?: string;
  area_affected_sqm?: number;
  population_affected?: number;
  duration_hours?: number;
  survey_data?: string;
  location_name: string;
  latitude: number;
  longitude: number;
  created_at: string;
  category?: CategoryResponse;
  municipality?: MunicipalityResponse;
  timeline: TimelineEventResponse[];
  attachments: AttachmentResponse[];
  assigned_officer?: string;
  assignedOfficer?: string;
  resolution?: ResolutionResponse;
}

export interface ComplaintCreatePayload {
  title: string;
  description: string;
  category_id: string;
  location_name: string;
  latitude: number;
  longitude: number;
  municipality_id?: string;
  area_affected_sqm?: number;
  population_affected?: number;
  duration_hours?: number;
  survey_data?: Record<string, unknown>;
}
